cherab/solps/formats/balance.py

```diff
@@ -93,65 +61,111 @@ def load_solps_from_balance(balance_filename):
         if fhandle.variables['species'].data[i, 1] == b'A' and fhandle.variables['species'].data[i, 2] == b'r':
             zn = 18
 
-        am = float(fhandle.variables['am'].data[i])  # Atomic mass
+        am = int(round(float(fhandle.variables['am'].data[i])))  # Atomic mass
         charge = int(fhandle.variables['za'].data[i])  # Ionisation/charge
-        species = _popular_species[(zn, am)]
+        isotope = lookup_isotope(zn, number=am)
+        species = prefer_element(isotope)  # Prefer Element over Isotope if the mass number is the same
 
-        # If we only need to populate species_list, there is probably a faster way to do this...		
-        sim.species_list.append(species.symbol + str(charge))
+        # If we only need to populate species_list, there is probably a faster way to do this...
+        species_list.append((species.name, charge))
 
-    tmp = fhandle.variables['na'].data.copy()
-    tmp = np.moveaxis(tmp, 0, -1)
-    sim._species_density = tmp
+    sim = SOLPSSimulation(mesh, species_list)
+
+    # TODO: add code to load SOLPS velocities and magnetic field from files
 
-    # Make Mesh Interpolator function for inside/outside mesh test.
-    inside_outside_data = np.ones(mesh.num_tris)
-    inside_outside = AxisymmetricMapper(Discrete2DMesh(mesh.vertex_coords, mesh.triangles, inside_outside_data, limit=False))
-    sim._inside_mesh = inside_outside
+    # Load electron species
+    sim.set_electron_temperature(fhandle.variables['te'].data.copy() / el_charge)
```

Comment:
ignore me it this needs to be optimised for speed.... but I suspect it doesn't need to be.

Reply:
There is a discussion in #20 and also in comments to the `SOLPSSimulation` code, that `sim.electron_temperature`, etc. should return not the data arrays but the interpolators. However, this is not decided yet. If `sim.electron_temperature` return the interpolator, than setting the electron temperature data in the following style:
```
sim.electron_temperature = data
```
will lead to the wizard hat paradox, because the user puts one object there, but then gets the different object.

Reply:
Probably the best thing would be to have different properties for data and interpolators, like
```
sim.electron_temperature_data  # for grid data
sim.electron_temperature  # for interpolator
```
@jacklovell, @Mateasek, what do you think?

Reply:
I would implement it as you suggest in the most recent comment. If users wish to obtain either an interpolator or the raw data, they should be accessible via separate attributes.

I am curious why these values are not being set via the __init__ rather than updating each attribute. Is there a plan for people to write data by creating an empty object first? If not it would be cleaner to configure the object via an init - like the Equilibrium class.

Reply:
Actually ignore the last point, it is just style and we use a similar pattern for the Plasma object.

Reply:
Ok I've just had a better look through the code to remind myself. I think there is some confusion here. The simulation object was just meant to be a thin wrapper holding the raw data. So all the properties should be data. The data gets converted to functions when you ask for a plasma object. @Mateasek was asking for additional methods / properties that return the data in functional form i.e. splitting out the code from the create_plasma() method. Can I suggest the following:

```
sim.electron_temperature                   # property, raw data arrays
sim.electron_temperature_as_f2d()    # method, function 2d
sim.electron_temperature_as_f3d ()   # method, function 3d
```

Reply:
I like this interface. I will replace the `set _... ()` methods with property setters and remove the `get _... ()` methods. I suggest a separate PR for adding the interpolators, because we have to decide whether we use `Discrete2DMesh` and `AxisymmetricMapper` or `SOLPSFunction3D` (see my comment in #20).

Reply:
I would use the ```SOLPSFunction3D``` for these. btw I've always been very uncomfortable with the use of the ```AxisymmetricMapper``` and ```Discrete2DMesh``` functions here. Using a float to hold an integer is dangerous, it will (with absolute certainty) result in the wrong data being indexed. I would rewrite those functions and place the code in the ```SOLPSFunction*``` classes. Have a quick look at the ```Discrete2DMesh``` code:
https://github.com/raysect/source/blob/development/raysect/core/math/function/float/function2d/interpolate/discrete2dmesh.pyx

It is trivial to get it to just return the index. See line 174. I would use the KDtree directly in the SOLPS function: 

https://github.com/raysect/source/blob/development/raysect/core/math/function/float/function2d/interpolate/common.pyx

Make the SOLPS functions 2D and them use the axisymmetric wrapper around the SOLPSFunction2D to convert it to 3D, when the result is actually a double!

Reply:
I implemented `SOLPSFunction2D` and `SOLPSVectorFunction2D` classes. Also added `*_f2d()` and `*_f3d()` interpolators for plasma parameters. The demos work, but more testing will probably be needed, which I'll leave for tomorrow.

Reply:
Nice work. I see you went with properties due to the low overhead of instancing the existing inside function. All looks good, much cleaner and more reliable.